### PR TITLE
Add final_layer_norm to OPT model

### LINKED ISSUE #17653: Abnormal behavior of OPT except OPT-350m
### System Info

```shell
- `transformers` version: 4.20.0.dev0
- Platform: Linux-5.11.0-1020-azure-x86_64-with-debian-bullseye-sid
- Python version: 3.7.13
- Huggingface_hub version: 0.7.0
- PyTorch version (GPU?): 1.10.1+cu111 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
```

### Who can help?

@patrickvonplaten @LysandreJik 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction
```
def log_probs_with_ppl(path, prompt):
    from transformers import AutoModelForCausalLM, AutoTokenizer
    import torch
    import torch.nn.functional as F
    # for half precision (13b models): torch_dtype=torch.float16
    model = AutoModelForCausalLM.from_pretrained(path).cuda()
    model.eval()
    tokenizer = AutoTokenizer.from_pretrained(path, use_fast=False)
    input_ids = tokenizer(prompt, return_tensors="pt").input_ids.cuda()
    outputs = model(input_ids, labels=input_ids)
    logits = outputs.logits
    arg_probs, _ = F.softmax(logits, dim=-1).max(-1)
    print("argmax probility:", arg_probs[0].cpu().detach().numpy())
    log_probs, tokens = F.log_softmax(logits, dim=-1).max(-1)
    print("argmax log probability:", log_probs[0].cpu().detach().numpy())
    sent = tokenizer.decode(tokens.squeeze().cpu().detach().numpy(), skip_special_tokens=False)
    print("argmax tokens:", sent)
    xentropy_loss = outputs[0]
    print("cross entropy loss:", xentropy_loss.item())
    ppl = torch.exp(xentropy_loss).item()
    print("ppl:", ppl)


if __name__ == "__main__":
    model_path = 'huggingface/opt-1.3b'
    prompts = "There is a book on the desk."
    log_probs_with_ppl(model_path, prompts)
```

### Expected behavior
The scripts above are used to test both the gpt2 and opt models, the results are shown below:

#### Sentence Scoring-GPT2
```
Input: There is a book on the desk.

# gpt2
argmax probility: [0.05637151 0.27859244 0.08230144 0.11579145 0.1521898  0.34015855 0.1640605  0.16998181]
argmax log probability: [-2.8757913 -1.2780054 -2.4973667 -2.1559646 -1.8826268 -1.0783434 -1.8075199 -1.772064 ]
argmax tokens:  is no lot called the subject of It
cross entropy loss: 4.143580913543701
ppl: 63.02811813354492


# gpt2-Medium
argmax probility: [0.39332193 0.26675868 0.08419792 0.1576896  0.2581378  0.1720277 0.1351828  0.13347614]
argmax log probability: [-0.93312687 -1.3214109  -2.474585   -1.8471267  -1.3542618  -1.7600996 -2.0011272  -2.0138326 ]
argmax tokens:  are no lot called the subject in
cross entropy loss: 3.641242504119873
ppl: 38.13919448852539

# gpt2-Large
argmax probility: [0.33576348 0.27546927 0.09161323 0.16216931 0.29808053 0.09624117 0.16370784 0.15139417]
argmax log probability: [-1.0913483 -1.2892792 -2.3901796 -1.8191143 -1.2103915 -2.340898 -1.809672  -1.8878684]
argmax tokens:  are no lot called the subject, It
cross entropy loss: 3.1841206550598145
ppl: 24.146047592163086
```

#### Sentence Scoring-OPT

```
Input: There is a book on the desk.

# opt-125m
argmax probility: [0.00063085 0.00046801 0.00079859 0.00062031 0.00056935 0.00048211 0.00078747 0.00045703 0.00154377]
argmax log probability: [-7.368442  -7.6670203 -7.1326685 -7.3852873 -7.471012  -7.6373434 -7.146683  -7.690766 -6.4735274]
argmax tokens: I aren nothing difference called youtube subjectawaru It
cross entropy loss: 9.004321098327637
ppl: 8138.173828125

# opt-350m
argmax probility: [0.09890626 0.29053134 0.30989376 0.05687688 0.31782693 0.24106818 0.15811151 0.12125468 0.2616786 ]
argmax log probability: [-2.313583  -1.2360439 -1.1715258 -2.8668664 -1.1462483 -1.4226755 -1.8444548 -2.109862 -1.3406382]
argmax tokens: 's a lot called the subject in
cross entropy loss: 3.714618682861328
ppl: 41.042930603027344

# opt-1.3b
argmax probility: [0.18575612 0.51934767 0.6326897  0.51414996 0.97731984 0.9402624 0.63661176 0.20046458 0.6865138 ]
argmax log probability: [-1.6833206  -0.65518177 -0.4577752  -0.6652403  -0.0229413  -0.06159634 -0.45159534 -1.6071177  -0.37612897]
argmax tokens: I are no difference called Amazon subject beside It
cross entropy loss: 7.282720565795898
ppl: 1454.94091796875

# opt-6.7b
argmax probility: [0.9414 0.391  0.766  0.627  0.998  0.7646 0.978  0.4473 0.8735]
argmax log probability: [-0.0602   -0.939    -0.2664   -0.4668   -0.001997 -0.268    -0.02211 -0.8047   -0.135   ]
argmax tokens: I's no lot called this subject. 
cross entropy loss: 7.17578125
ppl: 1307.0
```

When the model size increases, gpt2 tends to predict more accurate results with smaller ppl. However, opt models (except opt-350m) produce much larger ppl than the ppl of opt-350m.

Besides, it is abnormal that when the model size increases, opt models seem to have larger confidence score about the argmax decoding tokens (check **argmax probility** above).

I wonder what is causing such an issue. Looking forward to your reply. Thx!

### PR DESCRIPTION
**Pull request by thomasw21**
# What does this PR do?

Fixes #17653 , #17545

OPT models have a final_layer_norm: https://github.com/facebookresearch/metaseq/blob/e0c4f6b0e4c523906ad8d561f727e3f2ac3a8e73/metaseq/models/transformer.py#L466-L477
So we update HF models + conversion script to take in account that missing layer norm.

Test on OPT-125m (`restored.pt` file from `patrickvonplaten/opt_metaseq_125m`):
```
>>> model_path="fixed_opt_125m"
>>> prompt="Hello my name is"
>>> log_probs_with_ppl(model_path, prompt)
Input torch.Size([1, 5])
Logits torch.Size([1, 5, 50272])
torch.return_types.max(
values=tensor([[0.2398, 0.2326, 0.3332, 0.9363, 0.0097]], grad_fn=<MaxBackward0>),
indices=tensor([[ 100,    6,  766,   16, 1236]]))
argmax probility: [[0.23982257 0.23258895 0.33315504 0.9362957  0.00967377]]
argmax log probability: [[-1.4278558  -1.4584825  -1.0991473  -0.06582398 -4.6383367 ]]
argmax tokens: I, name is j
cross entropy loss: 4.051314830780029
ppl: 57.47297286987305
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
LGTM we should update the `tf` and flax checkpoints

**Comment by younesbelkada** [comment]
Thank you very much for the fix! I think that we'll have to change the generation tests a bit for the other models as well

**Review by younesbelkada — APPROVED** [review]

**Comment by ArthurZucker** [comment]
@thomasw21  I can update the tests and check the outputs if you want

**Comment by thomasw21** [comment]
@patrickvonplaten from what I understood logits comparison equality test were only done in 350m? @younesbelkada 
I actually manually converted `restored.pt` from https://huggingface.co/patrickvonplaten/opt_metaseq_125m using the updated conversion script.

@ArthurZucker thanks!

File: src/transformers/models/opt/convert_opt_original_pytorch_checkpoint_to_pytorch.py [hunk 1]
```diff
@@ -37,8 +37,6 @@ def load_checkpoint(checkpoint_path):
     # pop unnecessary weights
     keys_to_delete = [
         "decoder.version",
-        "decoder.layer_norm.weight",
```
Comment (patrickvonplaten) [inline src/transformers/models/opt/convert_opt_original_pytorch_checkpoint_to_pytorch.py:40]: That was the problem I think why we didn't notice that weights were missing

Comment (patrickvonplaten) [inline src/transformers/models/opt/convert_opt_original_pytorch_checkpoint_to_pytorch.py:40]: That was the problem I think why we didn't notice that weights were missing

File: src/transformers/models/opt/modeling_opt.py [hunk 2]
```diff
@@ -492,7 +492,11 @@ def __init__(self, config: OPTConfig):
         else:
             self.project_in = None
 
-        self.layer_norm = None
+        if config.do_layer_norm_before:
```
Comment (patrickvonplaten) [inline src/transformers/models/opt/modeling_opt.py:495]:
```suggestion
        if config.do_layer_norm_before and config.transformers_version >= 4.20.1:
```
(with correct regex)

Comment (thomasw21) [reply src/transformers/models/opt/modeling_opt.py:495]: Hum I think we'd much rather use a config instead ...

Comment (patrickvonplaten) [inline src/transformers/models/opt/modeling_opt.py:495]:
```suggestion
        if config.do_layer_norm_before and not config._remove_final_layer_norm:
```

File: src/transformers/models/opt/modeling_flax_opt.py [hunk 3]
```diff
@@ -452,6 +452,11 @@ def setup(self):
             self.project_in = None
             self.project_out = None
 
+        if self.config.do_layer_norm_before and not self.config._remove_final_layer_norm:
```
Comment (patrickvonplaten) [inline src/transformers/models/opt/modeling_flax_opt.py:455]:
```suggestion
        # Note that the only purpose of `config._remove_final_layer_norm` is to keep backward compatibility
        # with checkpoints that have been fine-tuned before transformers v4.20.1
        # see https://github.com/facebookresearch/metaseq/pull/164
        if self.config.do_layer_norm_before and not self.config._remove_final_layer_norm:
```

File: src/transformers/models/opt/modeling_tf_opt.py [hunk 4]
```diff
@@ -506,6 +506,11 @@ def __init__(self, config: OPTConfig, load_weight_prefix=None, **kwargs):
             name="embed_positions",
         )
 
+        if config.do_layer_norm_before and not config._remove_final_layer_norm:
```
Comment (patrickvonplaten) [inline src/transformers/models/opt/modeling_tf_opt.py:509]:
```suggestion
        # Note that the only purpose of `config._remove_final_layer_norm` is to keep backward compatibility
        # with checkpoints that have been fine-tuned before transformers v4.20.1
        # see https://github.com/facebookresearch/metaseq/pull/164
        if config.do_layer_norm_before and not config._remove_final_layer_norm:
```

File: src/transformers/models/opt/modeling_opt.py [hunk 5]
```diff
@@ -492,7 +492,11 @@ def __init__(self, config: OPTConfig):
         else:
             self.project_in = None
 
-        self.layer_norm = None
+        if config.do_layer_norm_before and not config._remove_final_layer_norm:
```
Comment (patrickvonplaten) [inline src/transformers/models/opt/modeling_opt.py:495]:
```suggestion
        # Note that the only purpose of `config._remove_final_layer_norm` is to keep backward compatibility
        # with checkpoints that have been fine-tuned before transformers v4.20.1
        # see https://github.com/facebookresearch/metaseq/pull/164
        if config.do_layer_norm_before and not config._remove_final_layer_norm:
```

**Review by patrickvonplaten — APPROVED** [review]
Thanks for fixing it! Let's merge the PR once the checkpoints are correctly uploaded under @ArthurZucker's namespace

File: src/transformers/models/opt/convert_opt_original_pytorch_checkpoint_to_pytorch.py [hunk 6]
```diff
@@ -37,8 +37,6 @@ def load_checkpoint(checkpoint_path):
...
-        "decoder.layer_norm.bias",
         "decoder.output_projection.weight",
```
Comment (younesbelkada) [inline src/transformers/models/opt/convert_opt_original_pytorch_checkpoint_to_pytorch.py:40]: This was ignored because in opt-125 and opt-350 the models use the input embedding layer to project it out to the vocabulary, see https://github.com/facebookresearch/metaseq/blob/e0c4f6b0e4c523906ad8d561f727e3f2ac3a8e73/metaseq/models/transformer.py#L486 
Since models can have different configs, I think it is better to double check that with the authors?

Comment (thomasw21) [reply src/transformers/models/opt/convert_opt_original_pytorch_checkpoint_to_pytorch.py:40]: Yep I sort of assumed that all models had that value to True. But indeed, didn't check besides 125M and 350M.

**Comment by younesbelkada** [comment]
When releasing the patch can we merge at the same time #17437 ? The problem of NaNs for batched generation still persists with this fix, but is resolved with #17437

**Comment by ArthurZucker** [comment]
BTW @patrickvonplaten do you have the expected values for the slow test?

**Comment by patrickvonplaten** [comment]
> BTW @patrickvonplaten do you have the expected values for the slow test?

Corrected the tests as well now

**Review by sgugger — APPROVED** [review]
Thanks for investigating, fixing and humoring my push for backward compatibility :-)
